File: experiments/X-6240-coefficient-stopping/chi_bound_fast.c
/* Log-space DP for Bmax(j) = max over qualifying length-j words of c_w/D.
   Every counterexample to chi = sigma with chi(n) = j has n <= Bmax(j).
   Tracking max log2(c) is exact for the max of c, since both transitions are monotone in c. */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

static double lae2(double a, double b){          /* log2(2^a + 2^b) */
  if(a < b){ double t=a; a=b; b=t; }
  if(a - b > 60) return a;
  return a + log2(1.0 + exp2(b - a));
}

int main(int argc, char **argv){
  int J = argc>1? atoi(argv[1]) : 30000;
  double V = argc>2? atof(argv[2]) : 2e9;
  const double L3 = log2(3.0), ALPHA = log(2.0)/log(3.0), NEG = -1e18;
  double *cur = malloc((J+3)*sizeof(double)), *nxt = malloc((J+3)*sizeof(double));
  for(int k=0;k<=J+2;k++) cur[k]=NEG;
  cur[0] = -1e15;                                 /* c = 0 at the root */
  double run = -INFINITY; int runj=0, cross=0;
  double marks[9]; int mk[9]={100,1000,5000,10000,15000,18000,20000,25000,30000};
  for(int i=0;i<9;i++) marks[i]=-INFINITY;
  for(int t=0;t<J;t++){
    int j=t+1;
    for(int k=0;k<=J+2;k++) nxt[k]=cur[k];        /* 0-step */
    for(int k=J+1;k>=0;k--){
      if(cur[k]<=NEG/2) continue;
      double v = lae2(cur[k]+L3, (double)t);      /* c -> 3c + 2^t */
      if(v > nxt[k+1]) nxt[k+1]=v;
    }
    int need = (int)ceil(ALPHA*j - 1e-12);
    double best = -INFINITY;
    for(int k=0;k<need && k<=J+1;k++){
      if(nxt[k]<=NEG/2) continue;
      double lk = k*L3;
      if(lk >= (double)j) continue;               /* need 2^j > 3^k */
      double logD = j + log2(1.0 - exp2(lk-(double)j));
      double r = nxt[k] - logD;
      if(r > best) best = r;
    }
    if(best > run){ run = best; runj = j; }
    if(!cross && run >= log2(V)) cross = j;
    for(int i=0;i<9;i++) if(j==mk[i]) marks[i]=run;
    for(int k=0;k<=J+2;k++) cur[k] = (k>=need)? nxt[k] : NEG;
  }
  printf("%7s %12s %14s\n","j<=","log2 Bmax","Bmax");
  for(int i=0;i<9;i++) if(mk[i]<=J && isfinite(marks[i]))
    printf("%7d %12.4f %14.4g\n", mk[i], marks[i], exp2(marks[i]));
  printf("\nV = %.4g  (log2 V = %.4f)\n", V, log2(V));
  if(cross){
    printf("running max first reaches V at j = %d\n", cross);
    printf("=> no counterexample to chi = sigma has chi(n) <= %d\n", cross-1);
  } else
    printf("running max never reaches V for j <= %d: max = %.6g at j = %d\n"
           "=> no counterexample to chi = sigma has chi(n) <= %d\n", J, exp2(run), runj, J);
  return 0;
}

File: experiments/X-6240-coefficient-stopping/chi_bound_fast.py
"""Log-space vectorised version of chi_bound.py, to locate where Bmax(j) crosses V.

State: for each k, the max over above-line prefixes of log2(c).  Transitions
  0-step: log2(c) unchanged, k unchanged
  1-step: c <- 3c + 2^t, i.e. log2 -> logaddexp2(log2(c)+log2(3), t)
Both are monotone in c, so tracking the max of log2(c) is exact for the max of c.
Bmax(j) = max over dropped k of  c/D  with log2(D) = j + log2(1 - 2^(k*log2 3 - j)).
"""
import numpy as np, math, sys

J = int(sys.argv[1]) if len(sys.argv) > 1 else 25000
V = 2e9
L3 = math.log2(3.0)
ALPHA = math.log(2)/math.log(3)
NEG = -1e18

logc = np.full(J+2, NEG)
logc[0] = NEG                     # c = 0 at the root; represent 0 as NEG
logc[0] = -1e18
cur = np.full(J+2, NEG); cur[0] = -1e18   # k index -> max log2(c)
cur[0] = -np.inf if False else -1e18

records = []
runmax = -np.inf; runj = None
for t in range(J):
    j = t+1
    # 1-step: from k to k+1 with c -> 3c + 2^t
    a = cur + L3
    one = np.logaddexp2(a, np.full_like(a, float(t)))
    nxt = cur.copy()                                   # 0-step keeps k
    nxt[1:] = np.maximum(nxt[1:], one[:-1])            # 1-step raises k by 1
    need = math.ceil(ALPHA*j - 1e-12)
    ks = np.arange(J+2)
    dropped = ks < need
    logD = j + np.log2(np.maximum(1.0 - np.exp2(np.minimum(ks*L3 - j, -1e-12)), 1e-300))
    ratio = np.where(dropped & (nxt > NEG/2) & (ks*L3 < j), nxt - logD, -np.inf)
    b = float(np.max(ratio)) if ratio.size else -np.inf
    records.append((j, b))
    if b > runmax:
        runmax, runj = b, j
    cur = np.where(ks >= need, nxt, NEG)

print(f"{'j':>7} {'log2 Bmax':>12} {'Bmax':>14}")
run = -np.inf
cross = None
for j, b in records:
    if b > run:
        run = b
        if j in (2,5,8,27,50,100) or (run > 10 and j % 1 == 0 and math.log2(V) - run < 40 and j % 500 == 0):
            pass
    if run >= math.log2(V) and cross is None:
        cross = j
for jm in (100, 1000, 5000, 10000, 15000, 18000, 20000, J):
    if jm <= J:
        r = max(b for j, b in records if j <= jm)
        print(f"{jm:>7} {r:>12.4f} {2**r:>14.4g}")
print(f"\nV = {V:.3g}  (log2 V = {math.log2(V):.4f})")
print(f"running max first reaches V at j = {cross}" if cross else
      f"running max never reaches V for j <= {J}: max = {2**run:.4g} at j = {runj}")
if cross:
    print(f"=> no counterexample to chi = sigma has chi(n) <= {cross-1}")
